8354520: IGV: dump contextual information

robcasloz opened the pull request:
This changeset extends the IGV graph dumps with additional properties that ease tracing the dumps back to the context in which they were produced. The changeset dumps, for every compilation, the following additional properties:

- JVM arguments
- platform information
- JVM version information
- date and time
- process ID
- (compiler) thread ID

![compilation-properties](https://github.com/user-attachments/assets/8ddc8fb9-c348-4761-8e19-c70633a1b59f)

Additionally, the changeset produces and dumps the C2 stack trace from which each graph is dumped:

![c2-stack-trace](https://github.com/user-attachments/assets/085547ee-b0b3-4a38-86f1-9df79cf1cc01)

This should be particularly useful in an interactive context, where the user steps through C2 code using a debugger and dumps graphs at different points. To produce a stack trace in this context, the usual debugger-entry C2 functions (`igv_print`, `igv_append`, `Node::dump_bfs`, ...) are extended with extra arguments to specify the stack handling registers (stack pointer, frame pointer, and program counter):

![c2-stack-trace-from-gdb](https://github.com/user-attachments/assets/29de2964-ee2d-4f5f-bcf7-d81e1bc6c8a6)

The inconvenience of manually specifying the stack handling registers can be addressed by hiding them in debugger user-defined commands, e.g.:

```
define igv
  p igv_print(true, $sp, $fp, $pc)
end

define igv_node
  p find_node($arg0)->dump_bfs(0, 0, "!", $sp, $fp, $pc)
end
```

Thanks to @TobiHartmann for providing useful feedback!

#### Testing

 - tier1 (windows-x64, linux-x64, linux-aarch64, macosx-x64, and macosx-aarch64; release and debug mode).
 - Tested interactive usage manually via `gdb` and `rr` on linux-x64.
 - Tested automatically that dumping thousands of graphs does not trigger any assertion failure.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8354520](https://bugs.openjdk.org/browse/JDK-8354520): IGV: dump contextual information (**Enhancement** - P4)


### Reviewers
 * [Damon Fenacci](https://openjdk.org/census#dfenacci) (@dafedafe - Committer) Review applies to [dd1ad6ad](https://git.openjdk.org/jdk/pull/24724/files/dd1ad6adccfaf132fa01e69bf54b394269a38741)
 * [Emanuel Peter](https://openjdk.org/census#epeter) (@eme64 - **Reviewer**)

### Reviewers without OpenJDK IDs
 * @mhaessig (no known openjdk.org user name / role) Review applies to [4d1e6f86](https://git.openjdk.org/jdk/pull/24724/files/4d1e6f867193759fb940984c590e41495f630ee3)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/24724/head:pull/24724` \
`$ git checkout pull/24724`

Update a local copy of the PR: \
`$ git checkout pull/24724` \
`$ git pull https://git.openjdk.org/jdk.git pull/24724/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 24724`

View PR using the GUI difftool: \
`$ git pr show -t 24724`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/24724.diff">https://git.openjdk.org/jdk/pull/24724.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/24724#issuecomment-2813092751)
</details>

mhaessig reviewed (approved):
Thank you for working on this and continually improving IGV!

I found two typos, but otherwise it looks good. I also tested printing to IGV from `rr` and the new features work as advertised.

Perhaps you could update the `dump_bfs` help with a comment about the new arguments?

mhaessig commented on src/utils/IdealGraphVisualizer/README.md:
```diff
@@ -48,6 +50,52 @@ Alternatively the output can be sent to a file using
+
+The JVM provides some entry functions to dump graphs from a debugger such as
+`gdb` or `rr`, see the different variants of `igv_print` and `igv_append` in
+`compile.cpp`. In combination with the IGV network interface, these functions
```
I often find them helpful to find the files a readme is referring to, so I tend to add them. But it's up to you whether you like them as well :slightly_smiling_face:

robcasloz replied:
Fair enough, so the relative link will resolve correctly and point to the actual `compile.cpp` file on GitHub?

robcasloz replied:
I added the link now (commit 8193767d). Turns out GitHub would treat it, by default, as relative to the directory where the `README.md` file is placed, so I had to prefix it with `/` to make it relative to the root of the repo.

robcasloz commented:
> I found two typos, but otherwise it looks good. I also tested printing to IGV from `rr` and the new features work as advertised.

Thanks for trying out and reviewing, Manuel!

> Perhaps you could update the `dump_bfs` help with a comment about the new arguments?

Done (commit d65ecdb5), let me know if that works for you.

I also made a minor correction to the IGV documentation (commit 86fdc673).

robcasloz commented:
Commit dd1ad6a documents how to dump the C2 stack trace when using `lldb` (default debugger on macOS platforms). Thanks @dafedafe for reporting and helping out!

dafedafe reviewed (approved):
Thanks a lot for adding this feature @robcasloz.

eme64 commented on src/hotspot/share/opto/compile.cpp:
```diff
@@ -5205,32 +5205,36 @@ IdealGraphPrinter* Compile::_debug_network_printer = nullptr;
 
 // Called from debugger. Prints method to the default file with the default phase name.
 // This works regardless of any Ideal Graph Visualizer flags set or not.
```
```suggestion
// This works regardless of any Ideal Graph Visualizer flags set or not.
// Use in debugger (gdb / rr): p igv_print($sp, $fp, $pc)
```

robcasloz replied:
Done (commit 6b8e659).

eme64 commented on src/hotspot/share/opto/compile.cpp:
```diff
@@ -5205,32 +5205,36 @@ IdealGraphPrinter* Compile::_debug_network_printer = nullptr;
 // the network flags for the Ideal Graph Visualizer, or to the default file depending on the 'network' argument.
 // This works regardless of any Ideal Graph Visualizer flags set or not.
-void igv_print(bool network) {
+void igv_print(bool network, void* sp, void* fp, void* pc) {
```
```suggestion
// Use in debugger (gdb / rr): p igv_print(true, $sp, $fp, $pc)
void igv_print(bool network, void* sp, void* fp, void* pc) {
```

robcasloz replied:
Done (commit 6b8e659).

eme64 commented on src/hotspot/share/opto/compile.cpp:
```diff
@@ -5205,32 +5205,36 @@ IdealGraphPrinter* Compile::_debug_network_printer = nullptr;
   }
 }
 
 // Same as igv_print(bool network) above but with a specified phase name.
```
```suggestion
// Same as igv_print(bool network, void* sp, void* fp, void* pc) above but with a specified phase name.
// Use in debugger (gdb / rr): p igv_print(true, "MyPhase", $sp, $fp, $pc)
```

robcasloz replied:
Done (commit 6b8e659).

eme64 commented on src/hotspot/share/opto/compile.cpp:
```diff
@@ -5242,40 +5246,42 @@ void igv_print_default() {
 // Called from debugger, especially when replaying a trace in which the program state cannot be altered like with rr replay.
 // A method is appended to an existing default file with the default phase name. This means that igv_append() must follow
 // an earlier igv_print(*) call which sets up the file. This works regardless of any Ideal Graph Visualizer flags set or not.
```
```suggestion
// an earlier igv_print(*) call which sets up the file. This works regardless of any Ideal Graph Visualizer flags set or not.
// Use in debugger (gdb / rr): p igv_append($sp, $fp, $pc)
```

robcasloz replied:
Done (commit 6b8e659).

eme64 commented on src/hotspot/share/opto/idealGraphPrinter.cpp:
```diff
@@ -336,6 +344,40 @@ void IdealGraphPrinter::begin_method() {
+
+  print_prop(COMPILATION_PROCESS_ID_PROPERTY, os::current_process_id());
+  print_prop(COMPILATION_THREAD_ID_PROPERTY, os::current_thread_id());
+
```
What about CPU features? Could be nice to know if we have `avx2` or `asimd`, etc.

robcasloz replied:
Done (commit cb781a95).

eme64 commented on src/hotspot/share/opto/idealGraphPrinter.cpp:
```diff
@@ -862,24 +904,88 @@ void IdealGraphPrinter::walk_nodes(Node* start, bool edges) {
 }
 
-void IdealGraphPrinter::print_graph(const char* name) {
+static bool skip_frame(const char* name) {
```
Nit: the name suggests that this is "skipping a frame". But you are asking if we "should skip the frame". So I would recommend a name like `is_skip_frame`, `must_skip_frame`, `should_skip_frame` or alike.

You could even invert the condition, and make the condition positive: `can_print_stack_frame`.

Totally optional, up to you :)

robcasloz replied:
Thanks, I went with `should_skip_frame` (commit 1210180b).

eme64 commented on src/hotspot/share/opto/idealGraphPrinter.cpp:
```diff
@@ -862,24 +904,88 @@ void IdealGraphPrinter::walk_nodes(Node* start, bool edges) {
+         strstr(name, "Node::dump_bfs") != nullptr;
+}
+
+static bool stop_frame_walk(const char* name) {
+  return strstr(name, "C2Compiler::compile_method") != nullptr;
+}
```
Nit: You could write `must_stop_frame_walk`. Btw, it could be nice to have a comment to explain why this is the condition to stop on. Maybe that comment could then turn into an even better method name?

robcasloz replied:
Thanks, I went with `should_end_stack_walk`, and added a comment as requested (commit 1210180b).

eme64 commented on src/hotspot/share/opto/idealGraphPrinter.cpp:
```diff
@@ -862,24 +904,88 @@ void IdealGraphPrinter::walk_nodes(Node* start, bool edges) {
+
+void IdealGraphPrinter::print_stack(frame fr, outputStream* graph_name) {
+  char buf[O_BUFLEN];
+  Thread* _current = Thread::current_or_null();
```
Is this a local variable? If so, I would drop the underscore. It generally suggests that it is a field, right?
It would also not hurt to call it `current_thread` for clarity.

robcasloz replied:
Thanks, the underscore was just a leftover from copying some code from `NativeStackPrinter::print_stack_from_frame()`. I just inlined `Thread::current_or_null()` into its only use, which I think is even clearer (commit `13604b75`).

robcasloz commented:
> Thanks a lot for adding this feature @robcasloz.

Thanks for reviewing, Damon!

eme64 reviewed (approved):
Nice work @robcasloz !

I just left a few suggestions below, but I think they are basically all nits / optional :)

robcasloz commented:
> Nice work @robcasloz !
> 
> I just left a few suggestions below, but I think they are basically all nits / optional :)

Thanks Emanuel, I addressed your comments and suggestions, including dumping of CPU features (as seen by C2, i.e. `Abstract_VM_Version::_features_string`).

robcasloz commented on src/hotspot/share/opto/compile.cpp:
```diff
@@ -5242,40 +5246,42 @@ void igv_print_default() {
+  Compile::current()->igv_print_method_to_file(nullptr, true, &fr);
 }
 
 // Same as igv_append() above but with a specified phase name.
```
Done (commit 6b8e659).

robcasloz commented on src/hotspot/share/opto/node.cpp:
```diff
@@ -2419,6 +2425,12 @@ void Node::dump_bfs(const int max_distance) const {
   dump_bfs(max_distance, nullptr, nullptr);
 }
 
+// Call this from debugger, with stack handling register arguments for IGV dumps.
```
Done (commit 6b8e659).

robcasloz commented on src/hotspot/share/opto/idealGraphPrinter.cpp:
```diff
@@ -862,24 +904,88 @@ void IdealGraphPrinter::walk_nodes(Node* start, bool edges) {
+  Thread* _current = Thread::current_or_null();
+  int count = 0;
+  int frame = 0;
+  while (count++ < StackPrintLimit && fr.pc() != nullptr) {
```
Done, thanks (commit 9cd331d2).

robcasloz commented on src/hotspot/share/opto/idealGraphPrinter.hpp:
```diff
@@ -110,6 +116,10 @@ class IdealGraphPrinter : public CHeapObj<mtCompiler> {
+  // Walk the native stack and print relevant C2 frames as IGV properties (if
+  // graph_name == nullptr) or the graph name based on the highest C2 frame (if
+  // graph_name != nullptr).
+  void print_stack(frame fr, outputStream* graph_name);
```
I refactored this code in commit 6fa328e0 to pass the initial frame by reference.

robcasloz commented on src/hotspot/share/opto/node.cpp:
```diff
@@ -1796,6 +1796,7 @@ class PrintBFS {
   const Node* _target;
   const char* _options;
   outputStream* _output;
+  frame* _frame;
```
Done (commit 3c3108b8), thanks.

eme64 reviewed (approved):
@robcasloz Amazing, thanks for the updates :)

robcasloz commented:
Manuel, Damon, and Emanuel: thanks again for reviewing!

robcasloz commented:
/integrate
